coderd/userauth.go

```diff
@@ -201,6 +202,237 @@ func (api *API) postConvertLoginType(rw http.ResponseWriter, r *http.Request) {
+		ctx               = r.Context()
+		auditor           = api.Auditor.Load()
+		logger            = api.Logger.Named(userAuthLoggerName)
+		aReq, commitAudit = audit.InitRequest[database.User](rw, &audit.RequestParams{
+			Audit:   *auditor,
+			Log:     api.Logger,
+			Request: r,
+			Action:  database.AuditActionWrite,
+		})
+	)
+	defer commitAudit()
+
+	if api.DeploymentValues.DisablePasswordAuth {
+		httpapi.Write(ctx, rw, http.StatusForbidden, codersdk.Response{
+			Message: "Password authentication is disabled.",
+		})
+		return
+	}
+
+	var req codersdk.ChangePasswordWithOneTimePasscodeRequest
+	if !httpapi.Read(ctx, rw, r, &req) {
+		return
+	}
+
+	//nolint:gocritic // In order to change a user's password, we need to get the user first!
+	user, err := api.Database.GetUserByEmailOrUsername(dbauthz.AsSystemRestricted(ctx), database.GetUserByEmailOrUsernameParams{
+		Email: req.Email,
+	})
+	if err != nil && !errors.Is(err, sql.ErrNoRows) {
+		logger.Error(ctx, "unable to fetch user by email", slog.Error(err))
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Internal error.",
+		})
+		return
+	}
+	aReq.Old = user
+
+	equal, err := userpassword.Compare(string(user.HashedOneTimePasscode), req.OneTimePasscode)
+	if err != nil {
+		logger.Error(ctx, "unable to compare passwords", slog.Error(err))
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Internal error.",
+		})
+		return
+	}
+
+	if !equal {
+		httpapi.Write(ctx, rw, http.StatusUnauthorized, codersdk.Response{
+			Message: "Incorrect email or one-time-passcode.",
+		})
+		return
+	}
+
+	if err := userpassword.Validate(req.Password); err != nil {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Invalid password.",
+			Validations: []codersdk.ValidationError{
+				{
+					Field:  "password",
+					Detail: err.Error(),
+				},
+			},
+		})
+		return
+	}
+
+	if equal, _ = userpassword.Compare(string(user.HashedPassword), req.Password); equal {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "New password cannot match old password.",
+		})
+		return
+	}
+
+	newHashedPassword, err := userpassword.Hash(req.Password)
+	if err != nil {
+		logger.Error(ctx, "unable to hash new user password", slog.Error(err))
+		httpapi.Write(ctx, rw, http.StatusInternalServerError, codersdk.Response{
+			Message: "Internal error hashing new password.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+
+	err = api.Database.InTx(func(tx database.Store) error {
+		//nolint:gocritic // We need to update the user's password.
+		err = tx.UpdateUserHashedPassword(dbauthz.AsSystemRestricted(ctx), database.UpdateUserHashedPasswordParams{
+			ID:             user.ID,
+			HashedPassword: []byte(newHashedPassword),
+		})
+		if err != nil {
+			return xerrors.Errorf("update user hashed password: %w", err)
+		}
+
+		//nolint:gocritic // We need to delete all API keys for the user.
```

Comment:
Why? API keys are separate authorization contexts to user passwords.

For example, as an operator I might've created an API key to use in CI to do `x`. It would be pretty weird and unexpected to invalidate that key if I forgot my personal password.

I'm not against this, but the reasons need to be carefully articulated **and** documented. When in doubt, follow the principle of least surprise.

Reply:
The `putUserPassword` endpoint has this behaviour. My assumption is that we'd want the behaviour to match. Unfortunately there are no comments in the immediate area justifying the decision.

https://github.com/coder/coder/blob/32a8df4b9b09c3aede8fd02a858a8167baaafb66/coderd/users.go#L977

https://github.com/coder/coder/blob/32a8df4b9b09c3aede8fd02a858a8167baaafb66/coderd/users.go#L1074-L1077

I'm happy to have a discussion around whether this should stay or go.

Reply:
Hhmm, do we document this behaviour anywhere?

Reply:
If we do I am unable to find it.

Reply:
Given that this behaviour exists, we can't change it now - and I think that it'd be the expectation of operators for this to behave similarly.

I _do_ think we should document this explicitly (could be done out-of-band of this PR), and raise an issue to clarify why we do this and possibly remove it for the reasons I outlined initially.

Reply:
@DanielleMaywood would you mind creating an issue for follow-up?